### ISSUE TITLE
Use aria-label in addition to text and title when matching?

### ISSUE THREAD

**Issue by lencioni** [jnicklas/capybara#1528]
Currently, with the following HTML:

```html
<button title="Click me">...</button>
```

this will work:

```ruby
expect(page).to have_button('Click me')
```

However, there are situations where you want to give elements text content for accessibility via `aria-label` and it would be useful if this content were matched as well.

```html
<button aria-label="Click me">...</button>
```

```ruby
expect(page).to have_button('Click me')
```

What do you think?

More information on `aria-label`: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute

**Comment by twalpole** [comment]
It's probably worthwhile to look at this for more than just buttons since aria-label can be applied to pretty much any html element.  This would require input from someone with a lot more knowledge of what makes sense from an aria perspective across all form elements, and probably for a v3 release.

**Comment by lencioni** [comment]
Yes, I agree that this should be used for more than just buttons--that was just the example I was using. Thanks!

**Comment by dnesteryuk** [comment]
I guess this request is related to [Xpath](https://github.com/jnicklas/xpath/) library which is used by Capybara.

**Comment by koenpunt** [comment]
To add support for `aria-label` to Capybara without waiting for the xpath library to implement it, you can just add this to your test helper:

```rb
module XPath::HTML

  # override link to support aria-label
  def link(locator)
    locator = locator.to_s
    link = descendant(:a)[attr(:href)]
    link[attr(:id).equals(locator) | string.n.is(locator) | attr(:'aria-label').is(locator) | attr(:title).is(locator) | descendant(:img)[attr(:alt).is(locator)]]
  end

end
```

**Comment by twalpole** [comment]
@koenpunt As of 2.7.0 Capybara no longer uses the XPath::HTML methods so your method won't work.  You'd need to modify (or replace) the :link and :button selectors  (Capybara.modify_selector/add_selector)

**Comment by koenpunt** [comment]
Oh well, I've missed that release. But still applies to older releases :) 

Why you've moved away from the XPath::HTML methods?

**Comment by twalpole** [comment]
@koenpunt Because they didn't allow the flexibility to make the locators optional
, and jnicklas is probably going to be removing XPath::HTML from the xpath gem

**Comment by twalpole** [comment]
PR #1732 Adds support for this that can be turned on by setting `Capybara.enable_aria_label = true` .  It defaults to `false` for backwards compatibility reasons.

**Comment by twalpole** [comment]
Closed via 22dfca46b8c8e9ddf272c44e9476a0687dc439cb
